### PR TITLE
add Lexer and lexing example

### PR DESCRIPTION
**Pull request by clayrat**
Closes https://github.com/gallais/idris-tparsec/issues/22

### CODE REVIEW

File: src/TParsec/Lexer.idr [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+-- a list of tokens. Some are associated to a token value (e.g. parentheses)
+-- others are not (e.g. space)
+  breaking : Char -> (b ** if b then Maybe Tok else ())
+-- Finally, strings which are not decoded as keywords are coerced using a
+-- function to token values.
+  default  : String -> Tok
+
+LexResult : {p : LexerParameters} -> Type
+LexResult {p} = List (Position, Tok p)
+
```
Comment (gallais) [inline src/TParsec/Lexer.idr]: Given that we need to pass `p` explicitly everywhere,
we may as well make it an explicit argument.

File: src/TParsec/Lexer.idr [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+-- We build a trie from the association list so that we may easily compute
+-- the successive derivatives obtained by eating the characters one by one
+Keywords : {p : LexerParameters} -> Type
+Keywords {p} = Trie Char (Tok p)
+
```
Comment (gallais) [inline src/TParsec/Lexer.idr]: The same is true for `Keywords`: we should just take `LexerParameters` as
an explicit argument.

**Comment by gallais** [comment]
More of a design discussion: I am not sure that insisting that the list of
keywords is non empty actually brings anything to the table. I am tempted to
drop this requirement when I finally update agdarsec to use tries.

**Comment by clayrat** [comment]
Yeah I was thinking about changing `keywords` to be a `SortedMap` instead.

**Comment by clayrat** [comment]
I've tried changing `keywords` to `SortedMap` but that creates problems for the typelevel test, because most of the declarations in that module are either `private` or `export`, so we can't evaluate the type completely. Also it doesn't buy us much since we only use it in `init` by immediately converting to `List`. OTOH a more specific type would guarantee key uniqueness, WDYT?

Also I've removed implicit declarations for `LexParameters`, but we'll have to bring most of them back for Idris2 to override the default erasure quantity.

**Comment by clayrat** [comment]
Btw, is there a reason to use `b ** if b then Maybe Tok else ()` instead of something like `Maybe (Maybe Tok)`?

**Comment by gallais** [comment]
I think `Maybe (Maybe Token)` blurs the intent. Tbh it may be worth introducing
a domain-specific sum type:

```agda
data MaybeBreaking : Set where
  IsBreaking : Maybe Tok -> MaybeBreaking
  IsNotBreaking : MaybeBreaking
```

**Comment by gallais** [comment]
> OTOH a more specific type would guarantee key uniqueness, WDYT?

Oh I see. I had not thought of that. An alternative is to use a fresh list, a cute
data structure defined by induction recursion for elements types with a decidable
equality.

**Comment by clayrat** [comment]
> An alternative is to use a fresh list, a cute
> data structure defined by induction recursion for elements types with a decidable
> equality.

Could you point me to some examples of fresh lists? Anyway I think this can probably be handled in a separate PR.

**Comment by gallais** [comment]
I have opened a PR to the Agda standard library with an implementation of fresh
lists together with documentation & an updated version of the lexer example using
fresh lists.

https://github.com/agda/agda-stdlib/pull/878

**Pull request merged**